crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -2384,6 +2384,259 @@ pub enum TryFromFilteredError {
     MissingWriteAllAccess,
 }
 
+/// A type-level set of components.
+///
+/// This is used to name components in [`EntityRefExcept`] and
+/// [`EntityMutExcept`].
+pub trait ComponentList {
+    /// An array of component IDs, one for each component in the list.
+    type Ids: ComponentIdList;
+
+    /// Returns the component IDs of each component in the list.
+    fn get_ids(components: &Components) -> Self::Ids;
+
+    /// Returns true if the component `C` is a member of this list and false
+    /// otherwise.
+    fn contains<C>() -> bool
+    where
+        C: Component;
+}
+
+/// A fixed-length array of component IDs.
+///
+/// This is essentially an implementation detail of [`ComponentList`].
+pub trait ComponentIdList: Clone + Send + Sync {
+    /// Returns true if this list contains the given component ID and false
+    /// otherwise.
+    fn contains(&self, id: ComponentId) -> bool;
+}
+
+impl<C> ComponentList for C
+where
+    C: Component,
+{
+    type Ids = ComponentId;
+
+    #[inline]
+    fn get_ids(components: &Components) -> Self::Ids {
+        components.component_id::<C>().unwrap()
+    }
+
+    #[inline]
+    fn contains<K>() -> bool
+    where
+        K: Component,
+    {
+        TypeId::of::<K>() == TypeId::of::<C>()
+    }
+}
+
+// Replaces a token tree with a constant. This is a helper for
+// `impl_component_list_tuple!` below.
+//
+// https://veykril.github.io/tlborm/decl-macros/patterns/repetition-replacement.html
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+
+// A helper macro to reduce internal boilerplate when for `ComponentList` tuple
+// implementations.
+macro_rules! impl_component_list_tuple {
+    ($($name: ident),*) => {
+        impl<$($name),*> ComponentList for ($($name,)*) where $($name: Component),* {
+            type Ids = [ComponentId; 0 $(+ replace_expr!($name 1))*];
+
+            #[inline]
+            fn get_ids(components: &Components) -> Self::Ids {
+                [$(components.component_id::<$name>().unwrap()),*]
+            }
+
+            #[inline]
+            fn contains<C>() -> bool where C: Component {
+                true $(|| TypeId::of::<C>() == TypeId::of::<$name>())*
```

Comment:
This will always return `true`, I guess you meant to use `false` here?

Reply:
I removed `ComponentList`.